### ISSUE TITLE
"Native" HTTP Client

### ISSUE THREAD

**Issue by vvcephei** [elastic/elasticsearch#7743]
My company (Bazaarvoice) is making heavy use of elasticsearch in our applications, and we've had a lot of trouble with the Java client options.

In short: both the Node and Transport clients suffer from sensitivity to differences in serialization. It's easy enough to maintain Java and ES version consistency within the cluster, but it's not so easy to do so between the cluster and client applications. Especially if you want to support operations like HA rolls. Using a JSON/REST client is much more permissive of variation between the client and server configurations.

On the other hand, the Java JSON client offerings have some pretty serious shortcomings of their own, mostly around having weird interfaces, but bugs are also a substantial challenge. 

Ideally, we would write all our code against the ES Client interface and then simply wire in an HttpTransportClient (e.g.). This would give the flexibility to experiment with the tradeoffs when using a Node client vs. a Transport client vs. Http. From a usablility perspective also, it seems preferable to reuse the same interfaces that ES already provides, documents, and maintains.

I've looked into the code, and it doesn't look like it would be too hairy, but I'd be surprised if you all hadn't thought about this problem at least once before.

I'd like to hear your thoughts before I get too deep into an implementation.
Also, I'd like to know if you'd have any interest in maintaining the library; specifically if you'd want it in the ES codebase or not. I can develop this in a separate repo, but if you think you'd ultimately want it, I should fork ES and send you a PR.

Thanks,
-John

**Comment by jtjeferreira** [comment]
hi @vvcephei 

you have any progress on your work? I also feel that the REST API support for the JVM story is not good. Currently we are using Jest and we also miss the aggregations support.

We are using ES via HTTP mainly due to authentication requirements.

**Comment by vvcephei** [comment]
Hey @jtjferreira

I had to put the project aside for longer than I wanted to focus on some
other major es projects at work, but I'm getting back into it for now. I'm
updating the dependency to ES 1.4.4, and I'm going to start with just get()
and index() to get a round-trip going. Once that is in place, I'll
publicize my repo and post a link to it here so everyone can get a more
concrete look at what I'm thinking.

I've already implemented these two functions on 0.90.7, so it shouldn't be
too long before I get to this phase. I got fairly bogged down before
working my way through search() because the facet response has so many
different forms. Hence, my desire to get _something_ out there.

Thanks for the interest. I won't leave you hanging too much longer.

-John

**Comment by vvcephei** [comment]
Ok, folks, I've sketched out what I've been talking about here: https://github.com/vvcephei/es-client-java

Please let me know what you think. For the long-term plans, see the README. For stuff that's on the roadmap, see the issues.

At this point, these are my top-level questions:
* What are the conditions under which ES adopt this project (once it's done) to be officially supported along with the other clients?
* Would ES at consider PRs to pull the fromXContent methods from this project to live next to the toXContent methods of the domain model objects? (thanks @kimchy  for that suggestion)
* Would ES consider PRs to serialize more information in the API (to allow us to reconstruct the Internal* domain objects from the json response)? I can see arguments on both sides of this, so I'd like to see what you think before I start sending PRs.

Thanks, everyone,
-John

**Comment by matt-blanchette** [comment]
+1 for same Client API but with HTTP transport
The Amazon ElasticSearch Service currently only supports HTTP

**Comment by davsclaus** [comment]
+1

**Comment by vvcephei** [comment]
Hey @matt-blanchette and @davsclaus,

It's good to hear I'm not the only one who thinks this should happen. You might want to check out the repo I previously linked (https://github.com/vvcephei/es-rest-client-java). Right now, it is only built against 1.3 and 1.4, but updating it is usually pretty straightforward. 

One of my coworkers did some pretty extensive functional testing of it and resolved a number of bugs. We didn't get a chance to to perf testing before we both got preoccupied with other concerns, though.

I haven't put much work into this project recently because:
* for us, the transport client is relatively safe now that they no longer use Java serialization.
* afiak, there's no concrete interest in having this library completed.

If you (or anyone else) want to use it, please let me know, and I'll make it a priority to support you.

Thanks,
-John

**Comment by s1monw** [comment]
FYI - we announced that we will build an official http client with minimal dependencies. I will assign this issue to me and will update it with the plans once we have them ready for you. I am pretty sure we will provide a basic version in the next couple of months but for now I just wanted to let everybody know we have it on the roadmap.

**Comment by vvcephei** [comment]
Oh, hey @s1monw! I missed that announcement. It's great to hear.

I don't know if it will help you, but feel free to mine my project if it helps you get started.

I'm also more than happy to volunteer if you want help on the official client.

Thanks,
-John

**Comment by otisg** [comment]
@s1monw this is still slated for 5.0.0?

**Comment by javanna** [comment]
> this is still slated for 5.0.0?

@otisg yes that is the goal.

**Comment by RichardWalshTZ** [comment]
Is there a JIRA issue number we can follow for this?

When you say basic API what do you mean, is there anywhere I can see a spec on what to expect.

Reason I ask is that I have hit the same limitations with the existing APIs that the other guys have experienced.

I am also now going to re-look at Spring Data ElasticSearch given that they have upgraded recently just to see what they have done.

**Comment by javanna** [comment]
> Is there a JIRA issue number we can follow for this?

We don't use JIRA but this is the right github issue for it. When there will be a pull request it will reference this issue.

> When you say basic API what do you mean, is there anywhere I can see a spec on what to expect.

Our REST spec are shared between all of our language clients and are here: https://github.com/elastic/elasticsearch/tree/master/rest-api-spec . The first version of the java client will be quite low level, with a very simple `performRequest` method that allows to send http requests to the cluster doing round-robin against multiple nodes and handling fail-over. We will then have api specific objects so that users can do things like `indexClient.index(index, type, id, body)` etc.

> Reason I ask is that I have hit the same limitations with the existing APIs that the other guys have experienced.

I think the biggest limitation is the fact that you have to depend on the whole elasticsearch to connect to it through java, plus bw compatibility on the transport layer, and those are the main reasons why we are working on the java REST client.

**Comment by gquintana** [comment]
It would be fantastic if this client could implement the current `Client` interface (like @vvcephei implementation) and be a drop in replacement for the `TransportClient` or `NodeClient`.

**Comment by javanna** [comment]
Hi @gquintana , I agree it would be nice, but I don't think it is likely to happen. I don't expect migrating to be particularly hard, but changes will be required. Also keep in mind that the very first version may be much more low level than the `Client` interface, which has calls for each specific api etc.

**Comment by robinst** [comment]
> We will then have api specific objects so that users can do things like indexClient.index(index, type, id, body) etc.

Is there another issue to follow the progress on this part?

**Comment by javanna** [comment]
Hi @robinst I created a meta issue for the planned improvements, which will be updated as we go: #19055 .
